src/closures/query_channel.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{z_closure_owned_query_drop, z_owned_closure_owned_query_t, z_owned_query_t};
+use libc::c_void;
+use std::sync::mpsc::TryRecvError;
+
+/// A closure is a structure that contains all the elements for stateful, memory-leak-free callbacks:
+/// - `this` is a pointer to an arbitrary state.
+/// - `call` is the typical callback function. `this` will be passed as its last argument.
+/// - `drop` allows the callback's state to be freed.
+///
+/// Closures are not guaranteed not to be called concurrently.
+///
+/// We guarantee that:
+/// - `call` will never be called once `drop` has started.
+/// - `drop` will only be called ONCE, and AFTER EVERY `call` has ended.
+/// - The two previous guarantees imply that `call` and `drop` are never called concurrently.
+#[repr(C)]
+pub struct z_owned_query_channel_closure_t {
+    context: *mut c_void,
+    call: Option<extern "C" fn(&mut z_owned_query_t, *mut c_void) -> bool>,
+    drop: Option<extern "C" fn(*mut c_void)>,
+}
+
+/// A pair of closures
+#[repr(C)]
+pub struct z_owned_query_channel_t {
+    pub send: z_owned_closure_owned_query_t,
+    pub recv: z_owned_query_channel_closure_t,
+}
+#[no_mangle]
+pub extern "C" fn z_query_channel_drop(channel: &mut z_owned_query_channel_t) {
+    z_closure_owned_query_drop(&mut channel.send);
+    z_query_channel_closure_drop(&mut channel.recv);
+}
+/// Constructs a null safe-to-drop value of 'z_owned_query_channel_t' type
+#[no_mangle]
+pub extern "C" fn z_query_channel_null() -> z_owned_query_channel_t {
+    z_owned_query_channel_t {
+        send: z_owned_closure_owned_query_t::empty(),
+        recv: z_owned_query_channel_closure_t::empty(),
+    }
+}
+
+/// Creates a new blocking fifo channel, returned as a pair of closures.
+///
+/// If `bound` is different from 0, that channel will be bound and apply back-pressure when full.
+///
+/// The `send` end should be passed as callback to a `z_get` call.
+///
+/// The `recv` end is a synchronous closure that will block until either a `z_owned_query_t` is available,
+/// which it will then return; or until the `send` closure is dropped and all queries have been consumed,
+/// at which point it will return an invalidated `z_owned_query_t`, and so will further calls.
+#[no_mangle]
+pub extern "C" fn zc_query_fifo_new(bound: usize) -> z_owned_query_channel_t {
+    if bound == 0 {
+        let (tx, rx) = std::sync::mpsc::channel();
```

Comment:
It seems a lot of code replication just for `std::sync::mpsc::channel()` vs `std::sync::mpsc::sync_channel(bound)` channel creation.

Reply:
kinda is, but since the types are distinct, the no-repeat alternative is generics, which would imply traitifying the channels, which is probably worse for now...

Reply:
Thanks for the explanation. It makes sense to me.